lib/html_cleaner.rb

```diff
@@ -220,153 +140,12 @@ def walk_array(array)
-      out_html += stack.add_p if put_inside_p_tag?(node.name) && !stack.inside_paragraph?
-
-      stack << [[node.name, Hash[*(node.attribute_nodes.map { |n| [n.name, n.value] }.flatten)]]]
-      out_html += open_tag(node)
-
-      # If we are the root node, pre-emptively open a paragraph
-      if node.name == "myroot"
-        out_html += stack.add_p
-      end
-
-      if no_break_before_after_tag?(node.name) and !stack.last.include?("p")
-        out_html += stack.add_p
-      end
-
-    else
-      text = node.to_s
-
-      # Remove leading/trailing linebreaks if we don't want to add
-      # additional linebreaks after the previous tag/before the next
-      # tag
-      prev_tag = node.previous_sibling.nil? ? "" : node.previous_sibling.name
-      text.lstrip! if no_break_before_after_tag?(prev_tag)
-      next_tag = node.next_sibling.nil? ? "" : node.next_sibling.name
-      text.rstrip! if no_break_before_after_tag?(next_tag)
-
-      out_html += stack.add_p if !stack.inside_paragraph? && text != ""
-      stack << [[node.name, Hash[*(node.attribute_nodes.map { |n| [n.name, n.value] }.flatten)]]]
-
-      # If we have three newlines, assume user wants a blank line
-      text.gsub!(/\n\s*?\n\s*?\n/, "\n\n&nbsp;\n\n")
-
-      # Convert double newlines into single paragraph break
-      text.gsub!(/\n+\s*?\n+/, stack.close_paragraph_tags + stack.open_paragraph_tags)
-
-      # Convert single newlines into br tags
-      text.gsub!(/\n/, '<br/>')
-
-      out_html += text
-    end
-
-    # decend into child nodes
-    node.children.each do |child|
-      stack, out_html = traverse_nodes(child, stack, out_html)
-    end
-
-    out_html += stack.close_and_pop_last
-    return [stack, out_html]
-  end
-
-
-  # Close an unclosed tag within the given text in the line at
-  # line_number, or before the next opening or closing tag if that
-  # comes first
-  def close_unclosed_tag(text, tag, line_number)
-    return text if self_closing_tag?(tag)
-    return text unless put_inside_p_tag?(tag)
-    line_number = line_number.to_i
-    lines = text.lines.to_a
-    pattern = /(^.*<#{tag}\s*.*?>.*?)($|<\/?\w+.*?\/?>)/
-    lines[line_number-1].gsub!(pattern, "\\1</#{tag}>\\2")
-    return lines.join("")
-  end
-
   def add_paragraphs_to_text(text)
-    doc = Nokogiri::XML.parse("<myroot>#{text}</myroot>")
-    doc.errors.each do |error|
-      match = error.message.match(/Opening and ending tag mismatch: (\w+) line (\d+) and myroot/)
-      text = close_unclosed_tag(text, match[1], match[2]) if match
-    end
-
     # Adding paragraphs in place of linebreaks
     doc = Nokogiri::HTML.fragment("<myroot>#{text}</myroot>")
-    out_html = traverse_nodes(doc.at_css("myroot"))[1]
-    # Remove empty paragraphs
-    out_html.gsub!(/<p>\s*?<\/p>/, "")
-    out_html.gsub!(/(\A<myroot>)|(<\/myroot>\Z)|(\A<myroot\/>\Z)/, "")
-    out_html
+    myroot = doc.children.first
+    ParagraphMaker.process(myroot)
+    myroot.children.to_xhtml
```

Comment:
Do you choose `to_xhtml` to be consistent with:

https://github.com/otwcode/otwarchive/blob/d74d404f052b4e60e223c712fff619844695806d/lib/html_cleaner.rb#L172-L174

or is there another reason?

Reply:
I chose `to_xhtml` because when I first wrote this with `to_s`, it introduced a test failure for this test:
https://github.com/otwcode/otwarchive/blob/d74d404f052b4e60e223c712fff619844695806d/features/other_a/parser.feature#L6-L21
The failure occurred not because it was parsing incorrectly, but because `to_s` adds a newline between the two characters when they're wrapped in a `<myroot>` tag:
```
2.7.3 :054 > Nokogiri::HTML.fragment("<myroot><p>1</p><p>2</p></myroot>").children.first.children.to_s
 => "<p>1</p>\n<p>2</p>" 
2.7.3 :055 > Nokogiri::HTML.fragment("<myroot><p>1</p><p>2</p></myroot>").children.first.children.to_xhtml
 => "<p>1</p><p>2</p>" 
```
Since `to_xhtml` matched the other code and didn't introduce the extra newline, it seemed like a reasonable choice to make. If you'd prefer I can switch to `to_s` (or some other output method) and fix up the test instead.

Reply:
I'm trying to see if relying on XHTML output is necessary, because we may eventually want to move away from it. This is related to @potpotkettle's comment below:

> Can you also fix [AO3-5801](https://otwarchive.atlassian.net/browse/AO3-5801)? I have #4303 but the approach here may be better.

The way this module works could probably be simplified even more. Right now, it does:

- `add_paragraphs_to_text` (which uses `Nokogiri::HTML.fragment`)
- `Sanitize.clean` (which [uses `Nokogiri::HTML5.fragment`](https://github.com/rgrove/sanitize/blob/v6.0.0/lib/sanitize.rb#L135-L144))
- then another round of `Nokogiri::HTML::Document.fragment`

If you remove the third part and avoid `to_xhtml`, it might help with AO3-5801.

But this change should be fine as is. We can tackle AO3-5801 separately.

Reply:
I agree that it might be good to eventually move away from XHTML, since I suspect it's responsible for https://otwarchive.atlassian.net/browse/AO3-4599 and possibly some other known issues. I think that XHTML is probably most useful for pretty formatting, because otherwise you'd get an unreadable blob of HTML tags in the Edit Work form, but that's also what the [textarea_convert_html_to_newlines.rb](https://github.com/otwcode/otwarchive/blob/master/config/initializers/monkeypatches/textarea_convert_html_to_newlines.rb) monkeypatch is for, so I could be completely off-base.

👍  for tackling AO3-5801 separately.